### ISSUE TITLE
WCF Memory Leak in 5.14 Pre-release

### ISSUE THREAD

**Issue by sonicmouse** [mono/mono#9207]
## Steps to Reproduce

run the attached solution:
[MemoryLeak.zip](https://github.com/mono/mono/files/2111628/MemoryLeak.zip)

On Windows, the 4 columns are `0; 0; 0; 0;`, on Lubuntu/Ubuntu they shoot up and the program eventually crashes due to the memory leak.

### Current Behavior

The 4 columns are not zero, representing the amount of bytes being leaked.

### Expected Behavior

The 4 columns should be zero.

## On which platforms did you notice this

[ ] macOS
[X] Linux
[ ] Windows

**Version Used**:
```
Mono JIT compiler version 5.14.0.125 (tarball Fri Jun  8 22:08:02 UTC 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           __thread
	SIGSEGV:       altstack
	Notifications: epoll
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          yes(3.6.0svn-mono-/)
	GC:            sgen (concurrent by default)
```

**Comment by sonicmouse** [comment]
We have a product going in to production in a week or so. This leak is bringing down all our Q/A machines within 5 hours. My manager is a tad worried.

Do we have an ETA on this? I told him if no movement is made on this issue in the next day, i will write my own "WCF"-like service.. just a simple TCP server that handles serialized RPCs. I don't want to do that, of course... but what other options are there in Mono? Anyone have any suggestions?

**Comment by paulfertser** [comment]
I would suggest you first somehow make sure you certainly have that fix of mine included. Then if it still leaks, you can try a similar transformation with ProcessInput function in the same file. Both shouldn't take much time, building a separate Mono instance seems to be trivial and well-documented. If that doesn't help, yes, you'll need to either restart your service every N requests (like PHP folks do :) ) or write something non-WCF-based.
Other than that, sorry, I'm not anyhow involved with the Mono project and I do not even know C#. Probably you can get private consulting from one of the Mono devs, @marek-safar , what do you think?

**Comment by sonicmouse** [comment]
The fix that was in your issue is definitely in 5.12+. I checked the merge.
.NET has a wonderful TCP socket implementation; Writing something trivial to replicate the the basic functionality of WCF shouldn't be too bad. Obviously I'd like to stick with the "tried and true" WCF framework instead of having to roll my own.

**Comment by paulfertser** [comment]
BTW, when you say the issue is not present with your testcase on "Windows" do you mean on windows running Mono or on Windows running Microsoft's .NET?

**Comment by sonicmouse** [comment]
In both .NET Framework and Mono, on Windows, they do not leak. I tested on both. Albeit it is incredibly slow in Mono, the columns all return 0 no matter what BCL it's running under on Windows.

This is strictly a Linux thing.

**Comment by marek-safar** [comment]
This is quite a big leak or memory issue which should make it easier to track it down.

**Comment by paulfertser** [comment]
@sonicmouse , I've checked https://www.mono-project.com/docs/about-mono/versioning/ and it says 5.14 version is kept in 2018-04 branch. My fix 62c0db64579b424fcd7b4de7fc8ec070643cc48d is certainly NOT present in it.
And the current "nightly" build is `Mono JIT compiler version 5.15.0.461 (tarball Sun May 20 11:40:32 UTC 2018)` so it doesn't include it either (just because it's very outdated).

**Comment by sonicmouse** [comment]
Doesn't look like this is a problem any longer. Tested on 5.18 of mono on ARM linux (debian) and Intel x64 linux (lubuntu)
